### PR TITLE
PARQUET-2292: Default SpecificRecord model reflects from MODEL$ field

### PR DESCRIPTION
**Pull request by clairemcginty**
Context:

AvroWriteSupport/AvroReadSupport can improve the precision of their default `model` selection. Currently they default to new SpecificDataSupplier().get()[0]. This means that SpecificRecord classes that contain logical types will fail out-of-the-box unless a specific DATA_SUPPLIER is configured that contains logical type conversions.

This PR attempts to derive a `model` by reflecting the `MODEL$` field from the SpecificRecordBase implementation, which in Avro 1.11 comes pre-loaded with all necessary logical type conversions for that Specific Avro class. (For Avro 1.8 compatibility, we have to add the conversions manually from the `conversions` field.)

### Jira

- [x] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (clairemcginty) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:179]: The reflection logic here is all based on the existing Avro behavior for its model cache: https://github.com/apache/avro/blob/release-1.11.1/lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java#L76-L86

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (clairemcginty) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:201]: For reference, Avro 1.11 compiles SpecificRecord implementations that look like:

```java
@org.apache.avro.specific.AvroGenerated
public class LogicalTypesTest extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  ...
  private static final SpecificData MODEL$ = new SpecificData();
  static {
    MODEL$.addLogicalTypeConversion(new org.apache.avro.data.TimeConversions.TimestampMillisConversion());
  }
}
```

Whereas for Avro 1.8, there is no static initialization, but there _is_ a `conversions` field:

```java
@org.apache.avro.specific.AvroGenerated
public class LogicalTypesTest extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  ...
  private static final SpecificData MODEL$ = new SpecificData();
  
  protected static final org.apache.avro.data.TimeConversions.TimestampConversion TIMESTAMP_CONVERSION = new org.apache.avro.data.TimeConversions.TimestampConversion();

  private static final org.apache.avro.Conversion<?>[] conversions =
      new org.apache.avro.Conversion<?>[] {
      TIMESTAMP_CONVERSION,
      null
  };
}
```

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java
Comment (wgtmac) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java:405]: What if `schema == null`?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroReadSupport.java
Comment (wgtmac) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroReadSupport.java:168]: Could you please revert these format changes?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (wgtmac) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:193]: Should it return if model is set successfully?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (wgtmac) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:179]: @gszadovszky Could you help review this? I am not that familiar with Avro.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java
Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java:405]: From where it's called in the code, `schema` should never be null -- but I guess it would be safest to add a null-check here to future-proof it a bit. Added!

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroReadSupport.java
Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroReadSupport.java]: done, sorry about that!

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (RustedBones) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:199]: This won't work for avro pre 1.10 where the implementation is not defined, and will return `null`
Here is a [trick](https://github.com/sbt/sbt-avro/blob/ec5f8aa064d0314a97a876bfe1d2d1bdb26dcbb8/src/main/scala/com/github/sbt/avro/SbtAvro.scala#L28) to access the avro version for previous artifact

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:193]: sorry, can you expand on this a bit?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (gszadovszky) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:194]: I think, we should log this exception. I would be also nice to use specific exceptions instead of catching everything. WDYT?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (gszadovszky) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:179]: TBH I do not have a strong opinion on any. I am fine with the current one if it works.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (gszadovszky) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:208]: Similar to the previous comment about logging and catching specific exceptions.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (gszadovszky) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:201]: Since we are using reflections on private members there are no compatibility guarantees. We shall be very careful here. What about avro versions prior to 1.8? Also, what if it breaks in the future? Will the related unit test fail for a future Avro releases (in case of upgrading the Avro version in the pom)?

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestSpecificReadWrite.java
Comment (gszadovszky) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestSpecificReadWrite.java:245]: We need to test both of `getModelForSchema` related to the avro version. If the version check gets more complicated, maybe more versions are to cover.

**Review by gszadovszky — CHANGES_REQUESTED** [review]
Sorry, @clairemcginty for the late response. I've added some comments.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:201]: yeah - it's unfortunate that this isn't solvable without reflection. I've manually tested this with earlier versions of Avro... let me see if I can reshape them into automated tests for `parquet-avro` 👀

Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java]: ok, I've added a new test suite, `TestAvroRecordConverter`, that thoroughly tests `getModelForSchema` with a variety of schema variants and with Avro versions 1.{7,8,9,10,11}. To do this, I used Powermock to mock the static invocation used to get Avro runtime version, then I ad-hoc compiled Avro generated classes using compiler versions 1.{7,8,9,10,11}. Happy to take a different approach with testing if you prefer -- this just seemed like the most straightforward way to do it :)

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java]: Added clearer logging & null-/error-handling.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java]: > Since we are using reflections on private members there are no compatibility guarantees. We shall be very careful here. What about avro versions prior to 1.8? Also, what if it breaks in the future? Will the related unit test fail for a future Avro releases (in case of upgrading the Avro version in the pom)?

so I've tested 1.7 and 1.8; since 1.9 Avro has stably used the `MODEL$` field to hold all conversions, so I feel reasonably confident about relying on this. If that changes, we'll catch it in the new unit tests 👍 

If you want, I can surround invocations of `getModelForSchema` in a try/catch (in `AvroReadSupport`/`AvroWriteSupport`), and just use the default SpecificDataSupplier if they throw anything. That way any unexpected behavior would just result in logical types not being used.

File: parquet-avro/pom.xml
Comment (wgtmac) [reply parquet-avro/pom.xml:116]: I think this is fine. BTW, is it compatible to upgrade this in the root pom as well?

**Review by wgtmac — COMMENTED** [review]
Thanks @gszadovszky for the review!

**Comment by wgtmac** [comment]
@clairemcginty Could you make the CIs happy?

File: parquet-avro/pom.xml
Comment (clairemcginty) [reply parquet-avro/pom.xml:116]: It should be, but it does require an artifact migration in a few modules (parquet-column and parquet-hadoop depend on mockito-all, which has moved to mockito-core in 2.x). I would be happy to do it, just not sure if it's in scope for this PR :)

**Comment by clairemcginty** [comment]
@wgtmac, sorry about that -- `mvn verify` was happy on my machine, but I wasn't running `install`. Should be good now!

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java
Comment (gszadovszky) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java]: > so I've tested 1.7 and 1.8; since 1.9 Avro has stably used the `MODEL$` field to hold all conversions, so I feel reasonably confident about relying on this. If that changes, we'll catch it in the new unit tests +1
This sounds perfect to me. Thanks a lot for the additional work!

> If you want, I can surround invocations of `getModelForSchema` in a try/catch (in `AvroReadSupport`/`AvroWriteSupport`), and just use the default SpecificDataSupplier if they throw anything. That way any unexpected behavior would just result in logical types not being used.
Yes, I think this fallback mechanism sounds reasonable to me.

Comment (clairemcginty) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java]: Great! I added code wrapping `getModelForSchema` in a try/catch at each call site.

**Review by gszadovszky — APPROVED** [review]
Thank you, @clairemcginty for your efforts. It looks good to me.

**Comment by clairemcginty** [comment]
thanks for the review @gszadovszky! very excited to start using this in Parquet :)

**Comment by gszadovszky** [comment]
I wouldn't like to make you sad, @clairemcginty, but we just released `1.13.0` last month and the previous one was almost a year ago. Meanwhile, nothing says we cannot do releases for tiny features, it is more about the effort and who has the time to do it. I usually don't, unfortunately.
Maybe if you ask nicely, @wgtmac would do another one... :wink:

**Comment by clairemcginty** [comment]
Oh no, I'm sorry I missed the boat! 😭 Is there anything that I (or my organization) could do to help out with the release, @wgtmac? If it helps at all, having this fix released would drastically help us drive Parquet adoption, and lead to more OSS contributions down the line :)

**Comment by wgtmac** [comment]
The parquet community is discussing a new `1.13.1` release to address some issues introduced by `1.13.0`: https://lists.apache.org/thread/1mjvdcmwqjcblmfkfgpd9ob2yodx7tom

As the release manager, what do you think if we include this PR to the `1.13.1` release? @Fokko

**Comment by gszadovszky** [comment]
@clairemcginty, could you reply on the thread that you would like to have this one included? It is better to have a broader audience for discussing the release content. (Note: We should usually not include features in patch releases to lower the potential risks. But it is up to the community to decide.)

**Comment by clairemcginty** [comment]
@gszadovszky I wasn't previously subscribed to the dev mailing list, so I just subscribed -- I think I need to wait until someone pings the thread for it to land in my inbox so I can reply :)

**Comment by gszadovszky** [comment]
@clairemcginty, I think, there is an option to reply from you own mail client in ponymail (the thread link above).

**Comment by clairemcginty** [comment]
thanks @gszadovszky! I used the Ponymail link from the thread, but I fear it's just created a new email thread - apologies

**Comment by Fokko** [comment]
Thanks for pining me here @wgtmac I just replied on the dev-list. I think we can include this one.

**Pull request merged**
